pkg/registry/core/service/ipallocator/ipallocator.go

```diff
@@ -0,0 +1,480 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipallocator
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"net/netip"
+	"time"
+
+	networkingv1alpha1 "k8s.io/api/networking/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	networkingv1alpha1informers "k8s.io/client-go/informers/networking/v1alpha1"
+	networkingv1alpha1client "k8s.io/client-go/kubernetes/typed/networking/v1alpha1"
+	networkingv1alpha1listers "k8s.io/client-go/listers/networking/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	netutils "k8s.io/utils/net"
+)
+
+const ControllerName = "ipallocator.k8s.io"
+
+// Allocator implements current ipallocator interface using IPAddress API object
```

Comment:
General Thoughts: 
My biggest concern is on the way we implemented this is it is o(N) where n is the # of IPs in the systems. That is quite something giving the fact that some systems will run with thousands if not 10s of thousands of IPs. the start from the top and from the bottom of the range helps but i am not sure of how good it is.

The second biggest concern I have is the idea that the expected stdev perf for a set of allocation operations is just too high. Think of the delta between  the fastest allocation (lucky to find a free ip early in the loop) or slowest (where we need to iterate over the entire list of IPs). 

if we agree that these are concern. Can we start by considering alternative options specifically replacing the iteration with a process that does not need to iterate over everything.

Reply:
fully agree, and I put a lot of thought on that and is also mentioned in the KEP, I also added benchmarks in both allocators to compare, see pkg/registry/core/service/ipallocator/ipallocator_test.go. and added it as a part of the beta criteria ;)

> Improve performance on the allocation logic, O(1) for allocating a free address

https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs#beta

But I think that for alpha this is good enough